### ISSUE TITLE
Add support for Imaginary to generate previews via "file" parameter through locally mounted folder with "-mount" command line option

### ISSUE THREAD

**Issue by invario** [nextcloud/server#44708]
The current implementation of Imaginary preview generation in Nextcloud works by submitting the image via a POST to the configured 'preview_imaginary_url' in config.php.

While this works and is scalable, I believe on certain (and possibly many) configurations, this is the slower route wherein the Imaginary server (whether running on baremetal, a jail, or Docker container) may have local file access to the image. Thus, having NC server send the entire image via a POST to Imaginary server is an unnecessary use of bandwidth and has network overhead, and is slower than local file access.

Imaginary currently has the ability to instead use a "file" parameter via the POST request to direct Imaginary to generate a preview for a file that it has local access to via the "-mount" parameter provided on the command line of Imaginary:
https://github.com/h2non/imaginary

> - file string - Use image from server local file path. In order to use this you must pass the -mount=<dir> flag.

**Describe the solution you'd like**
A boolean option in config.php such as 'preview_imaginary_local_mount' that would direct NC to instead send a proper file path via "file" and not upload the entire file to Imaginary.  This would, of course, necessitate that the instance of Imaginary being used is configured properly with the NC Data folder mounted.

**Describe alternatives you've considered**
ImageMagick previews are currently generated with direct access to the files due to the fact that ImageMagick runs on the same server via a PHP Module.  I recently switched over to Imaginary because of 1) webp previews being possible and 2) slight security issues in ImageMagick (none that I've encountered, but I've been urged to jump to Imaginary.) I would sleep better at night (lol) knowing that Imaginary is running as fast as or faster than ImageMagick for preview generation but as long as NC has to send an entire file via POST to Imaginary, I think that would cause a large performance hit via a locally access ImageMagick generated preview.

If I have some time, I may run some tests with say 100 random images in my NC instance and use PreviewGenerator to preview:generate-all those 100 images with Imaginary and ImageMagick for comparison.

Thank you for your time if you read this.

**Comment by invario** [comment]
My first time working on something like this. It appears to be working on my end.

**Comment by kesselb** [comment]
> If I have some time, I may run some tests with say 100 random images in my NC instance and use PreviewGenerator to preview:generate-all those 100 images with Imaginary and ImageMagick for comparison.

Please :pray:

**Comment by joshtrichards** [comment]
Cc: @szaimen Something to evaluate for standalone AIO maybe?

**Comment by invario** [comment]
> > If I have some time, I may run some tests with say 100 random images in my NC instance and use PreviewGenerator to preview:generate-all those 100 images with Imaginary and ImageMagick for comparison.
> 
> Please 🙏

Ha, well, this is a tiny bit surprising but I ran a few tests and here's what I discovered. On *my* specific setup, there appears to be absolutely no gains from the change I made to use locally mounted files as opposed to using the existing POST method with the entire file being uploaded.

I also took the liberty of comparing ImageMagick and Imaginary is **a lot faster** both with local file access and via POST, at least when generating JPG. When generating webp, the performance is the same as ImageMagick.

I did my best to control all the variables including server load. I don't have many/any regularly active users so it wasn't terribly difficult. I ran each test case 3 times using:
`time php occ preview:generate-all --path=/user/test/testfiles`

In between runs, I deleted the "/user/test/testfiles" folder using NC Files app and for good measure executed a "php occ files:scan --path=/user/test" to ensure all was wiped.  Then I recreated the folder, copied the same files in (via SSH), re-executed "php occ files:scan --path=/user/test", then finally reran the test command. 


**Test setup:**
Ryzen 5 1600 AF (6 core, 12 threads)
TrueNAS Core 13 with NC 28.0.4 running in iocage jail (FreeBSD 13.2-RELEASE)
175 HEIC files, resolutions of 4000x3000 taken by my Samsung GS23+
Total directory size: 918,952 KB
innodb_flush_log_at_trx_commit=0 (best performance in MariaDB/MySQL for development)
sync_binlog=0 (best performance in MariaDB/MySQL for development)
Imaginary (version 1.2.4)
vips 8.15.1
preview squaresizes=64
preview widthsizes=256 1024 1920
preview lengthsizes=256 1024 1920
preview_max_x=2048
preview_max_y=2048
webp_quality=80
jpeg_quality=60
cropimagepreviews=0

**Imaginary - locally_mounted=1, webp**
4:28.34
4:19.31
4:27.46
Avg: 4:25

**Imaginary - locally_mounted=0, webp**
4:28.16
4:23.05
4:18.88
Avg: 4:23

**ImageMagick 7.1.1-26 Q8 (no HDRI), jpg**
4:27.02
4:33.87
4:33.27
Avg: 4.31

**Imaginary - locally_mounted=1, jpg**
2:51.75
2:48.91
2:43.57
Avg: 2:48

**Imaginary - locally_mounted=0, jpg**
2:44.32
2:56.93
2:54:09
Avg: 2:52

**Tentative Conclusions:**

1. Imaginary for generating JPG is definitely faster than Imagemagick, regardless of local file mount or via POSTS.
2. Imaginary for generating WEBP runs as fast as or slightly faster than Imagemagick Q8 (non-HDRI) generating JPG.
3. Most pre-compiled ImageMagick distributions are Q16-HDRI, so that will be even slower, thereby increasing the performance gap in favor of Imaginary.
4. My specific setup uses large HEIC files so JPG and/or smaller files will undoubtedly perform differently... possibly making local file processing more important?
5. Generating previews by using "occ preview:generate" is a single-threaded process. I do not know how multiple requests sent to Imaginary or ImageMagick would perform.
